pauliopt/clifford/tableau.py

```diff
@@ -0,0 +1,381 @@
+        ct = CliffordTableau(n_qubits)
+        ct.tableau = qiskit_ct.symplectic_matrix
+        ct.signs = qiskit_ct.phase
+        return ct
+
+    def string_repr(self, sep=" ", sign_sep="| "):
+        """
+        Get a string representation of the tableau.
+
+        Args:
+            sep (str): Separator between the pauli operators
+            sign_sep (str): Separator between operators and sign.
+
+        """
+        out = ""
+        for i in range(self.n_qubits):
+            for j in range(self.n_qubits):
+                x_str = ["I", "X", "Z", "Y"][int(self.x_out(i, j))]
+                z_str = ["I", "X", "Z", "Y"][int(self.z_out(i, j))]
+                out += f"{x_str}/{z_str}" + sep
+            out += sign_sep + f"{'+' if self.signs[i] == 0 else '-'} \n"
+        return out
+
+    def x_out(self, row, col):
+        """
+        Get the X operator in row `row` and column `col`.
+
+        Args:
+            row (int): Row index.
+            col (int): Column index.
+        """
+        return self.tableau[row, col] + \
+               2 * self.tableau[row, col + self.n_qubits]
+
+    def z_out(self, row, col):
+        """
+        Get the Z operator in row `row` and column `col`.
+
+        Args:
+            row (int): Row index.
+            col (int): Column index.
+        """
+        return self.tableau[row + self.n_qubits, col] + \
+               2 * self.tableau[row + self.n_qubits, col + self.n_qubits]
+
+    def prepend_h(self, qubit):
+        """
+        Prepend a Hadamard gate to the tableau.
+
+        Args:
+            qubit (int): Qubit the hadamard gate is applied to.
+        """
+        self.signs[[qubit, self.n_qubits + qubit]] = self.signs[
+            [self.n_qubits + qubit, qubit]]
+        self.tableau[[self.n_qubits + qubit, qubit], :] = \
+            self.tableau[[qubit, self.n_qubits + qubit], :]
+
+    def append_h(self, qubit):
+        """
+        Append a Hadamard gate to the tableau.
+
+        Args:
+            qubit (int): Qubit the hadamard gate is applied to.
+        """
+        self.signs = (self.signs + self.tableau[:, qubit] * self.tableau[:,
+                                                            self.n_qubits + qubit]) % 2
+
+        self.tableau[:, [self.n_qubits + qubit, qubit]] = self.tableau[:,
+                                                          [qubit, self.n_qubits + qubit]]
+
+    def prepend_s(self, qubit):
+        """
+        Prepend a S gate to the tableau.
+
+        Args:
+            qubit (int): Qubit the S gate is applied to.
+        """
+        stabilizer = self.tableau[qubit, :]
+        destabilizer = self.tableau[qubit + self.n_qubits, :]
+        stab_sign = self.signs[qubit]
+        destab_sign = self.signs[qubit + self.n_qubits]
+
+        destabilizer, destab_sign = \
+            mult_paulis(stabilizer, destabilizer, stab_sign, destab_sign, self.n_qubits)
+        self.insert_pauli_row(destabilizer, destab_sign, qubit)
```

Comment:
Since every `mult_pauli` is followed by an `insert_pauli_row`, lets just make `xor_row(i, j)` a method in the tableau.